ICE: `failed to reparse`

<!--
[31mICE[0m: Rustc ./a.rs '' 'thread 'rustc' panicked at compiler/rustc_parse/src/parser/mod.rs:768:31: 'failed to reparse {mv_kind:?}: Some(DiagInner { level: Error, messages: [(FluentIdentifier("parse_expected_identifier_found_str", None), NoStyle)], code: None, span: MultiSpan { primary_spans: [/tmp/im/a.rs:9:17: 9:28 (#0)], span_labels: [(/tmp/im/a.rs:9:17: 9:28 (#0), FluentIdentifier("parse_expected_identifier", None))] }, children: [Subdiag { level: Help, messages: [(FluentIdentifier("parse_invalid_identifier_with_leading_number", None), NoStyle)], span: MultiSpan { primary_spans: [/tmp/im/a.rs:9:17: 9:25 (#0)], span_labels: [] } }], suggestions: Enabled([]), args: {"token": Str("0b101010f64")}, sort_span: /tmp/im/a.rs:9:17: 9:28 (#0), is_lint: None, long_ty_path: None, emitted_at: DiagLocation { file: "compiler/rustc_parse/src/parser/diagnostics.rs", line: 371, col: 34 } })'', 'thread 'rustc' panicked at compiler/rustc_parse/src/parser/mod.rs:768:31: 'failed to reparse {mv_kind:?}: Some(DiagInner { level: Error, messages: [(FluentIdentifier("parse_expected_identifier_found_str", None), NoStyle)], code: None, span: MultiSpan { primary_spans: [/tmp/im/a.rs:9:17: 9:28 (#0)], span_labels: [(/tmp/im/a.rs:9:17: 9:28 (#0), FluentIdentifier("parse_expected_identifier", None))] }, children: [Subdiag { level: Help, messages: [(FluentIdentifier("parse_invalid_identifier_with_leading_number", None), NoStyle)], span: MultiSpan { primary_spans: [/tmp/im/a.rs:9:17: 9:25 (#0)], span_labels: [] } }], suggestions: Enabled([]), args: {"token": Str("0b101010f64")}, sort_span: /tmp/im/a.rs:9:17: 9:28 (#0), is_lint: None, long_ty_path: None, emitted_at: DiagLocation { file: "compiler/rustc_parse/src/parser/diagnostics.rs", line: 371, col: 34 } })''
File: /tmp/im/a.rs
-->

auto-reduced (treereduce-rust):
````rust
pub fn main() {
    match X {
        deref!([0b101010f64, _, 3]) => {}
    }
}

````

original:
````rust
//@ run-pass

struct Foo;

static X: Foo = Foo;

pub fn main() {
    match X {
        deref!([0b101010f64, _, 3]) => {}
    }
}

````

Version information
````
rustc 1.87.0-nightly (351686bcf 2025-03-01)
binary: rustc
commit-hash: 351686bcfd18dd0f652aba69a806bfa68c57234d
commit-date: 2025-03-01
host: x86_64-unknown-linux-gnu
release: 1.87.0-nightly
LLVM version: 20.1.0

````

Possibly related line of code:
https://github.com/rust-lang/rust/blob/351686bcfd18dd0f652aba69a806bfa68c57234d/compiler/rustc_parse/src/parser/mod.rs#L762-L774

Command:
`/home/matthias/.rustup/toolchains/master/bin/rustc `

<details><summary><strong>Program output</strong></summary>
<p>

```
error[E0658]: use of unstable library feature `deref_patterns`: placeholder syntax for deref patterns
 --> /tmp/icemaker_global_tempdir.wxhLJkXfbxMZ/rustc_testrunner_tmpdir_reporting.FgtN60LYeTIA/mvce.rs:3:9
  |
3 |         deref!([0b101010f64, _, 3]) => {}
  |         ^^^^^
  |
  = note: see issue #87121 <https://github.com/rust-lang/rust/issues/87121> for more information
  = help: add `#![feature(deref_patterns)]` to the crate attributes to enable
  = note: this compiler was built on 2025-03-01; consider upgrading it if it is out of date


thread 'rustc' panicked at compiler/rustc_parse/src/parser/mod.rs:768:31:
failed to reparse {mv_kind:?}: Some(DiagInner { level: Error, messages: [(FluentIdentifier("parse_expected_identifier_found_str", None), NoStyle)], code: None, span: MultiSpan { primary_spans: [/tmp/icemaker_global_tempdir.wxhLJkXfbxMZ/rustc_testrunner_tmpdir_reporting.FgtN60LYeTIA/mvce.rs:3:17: 3:28 (#0)], span_labels: [(/tmp/icemaker_global_tempdir.wxhLJkXfbxMZ/rustc_testrunner_tmpdir_reporting.FgtN60LYeTIA/mvce.rs:3:17: 3:28 (#0), FluentIdentifier("parse_expected_identifier", None))] }, children: [Subdiag { level: Help, messages: [(FluentIdentifier("parse_invalid_identifier_with_leading_number", None), NoStyle)], span: MultiSpan { primary_spans: [/tmp/icemaker_global_tempdir.wxhLJkXfbxMZ/rustc_testrunner_tmpdir_reporting.FgtN60LYeTIA/mvce.rs:3:17: 3:25 (#0)], span_labels: [] } }], suggestions: Enabled([]), args: {"token": Str("0b101010f64")}, sort_span: /tmp/icemaker_global_tempdir.wxhLJkXfbxMZ/rustc_testrunner_tmpdir_reporting.FgtN60LYeTIA/mvce.rs:3:17: 3:28 (#0), is_lint: None, long_ty_path: None, emitted_at: DiagLocation { file: "compiler/rustc_parse/src/parser/diagnostics.rs", line: 371, col: 34 } })
stack backtrace:
   0:     0x7b7181dbf0d4 - <std::sys::backtrace::BacktraceLock::print::DisplayBacktrace as core::fmt::Display>::fmt::h512fc0ab65645de5
   1:     0x7b71826057e2 - core::fmt::write::hdbee15e0a2df9fd7
   2:     0x7b71839f6411 - std::io::Write::write_fmt::hc06675d21e69e541
   3:     0x7b7181dbef32 - std::sys::backtrace::BacktraceLock::print::h71b425936fc3e4b8
   4:     0x7b7181dc1812 - std::panicking::default_hook::{{closure}}::hbba6d1889ba06809
   5:     0x7b7181dc1404 - std::panicking::default_hook::ha1628e5560225566
   6:     0x7b7180f19137 - std[73fc39335cd1ceda]::panicking::update_hook::<alloc[7449611bcb229259]::boxed::Box<rustc_driver_impl[c7033648bfdf4c80]::install_ice_hook::{closure#1}>>::{closure#0}
   7:     0x7b7181dc2083 - std::panicking::rust_panic_with_hook::h9f3cc492432ffe0a
   8:     0x7b7181dc1d7a - std::panicking::begin_panic_handler::{{closure}}::h7255667d7ab4e1d4
   9:     0x7b7181dbf5a9 - std::sys::backtrace::__rust_end_short_backtrace::hcf69635c7990946c
  10:     0x7b7181dc1a3d - rust_begin_unwind
  11:     0x7b717ea5b2c0 - core::panicking::panic_fmt::ha881f14dc09ed01c
  12:     0x7b717ebe96b6 - core::result::unwrap_failed::h08d9e0d600154a82
  13:     0x7b7182f1f2a6 - <rustc_parse[3959de2b5713e24e]::parser::Parser>::parse_pat_with_range_pat
  14:     0x7b7182effaea - <rustc_parse[3959de2b5713e24e]::parser::Parser>::parse_pat_no_top_guard_inner
  15:     0x7b7182eff86a - <rustc_parse[3959de2b5713e24e]::parser::Parser>::parse_pat_allow_top_guard
  16:     0x7b71817734d7 - <rustc_parse[3959de2b5713e24e]::parser::Parser>::parse_builtin::<rustc_ast[fbe4be00de71ee9f]::ast::PatKind, <rustc_parse[3959de2b5713e24e]::parser::Parser>::parse_pat_builtin::{closure#0}>
  17:     0x7b7182f1ea6c - <rustc_parse[3959de2b5713e24e]::parser::Parser>::parse_pat_with_range_pat
  18:     0x7b7182effaea - <rustc_parse[3959de2b5713e24e]::parser::Parser>::parse_pat_no_top_guard_inner
  19:     0x7b7182eff86a - <rustc_parse[3959de2b5713e24e]::parser::Parser>::parse_pat_allow_top_guard
  20:     0x7b7182f53b0b - <rustc_expand[a7ac4f37f407a022]::mbe::macro_rules::ParserAnyMacro>::make
  21:     0x7b7180fcbe1d - <rustc_expand[a7ac4f37f407a022]::mbe::macro_rules::ParserAnyMacro as rustc_expand[a7ac4f37f407a022]::base::MacResult>::make_pat
  22:     0x7b7183931749 - <rustc_expand[a7ac4f37f407a022]::expand::MacroExpander>::fully_expand_fragment
  23:     0x7b718390656b - <rustc_expand[a7ac4f37f407a022]::expand::MacroExpander>::expand_crate
  24:     0x7b71829e0133 - rustc_interface[b107db532c72b558]::passes::resolver_for_lowering_raw
  25:     0x7b71829df58d - rustc_query_impl[ad104e1204eb17fd]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[ad104e1204eb17fd]::query_impl::resolver_for_lowering_raw::dynamic_query::{closure#2}::{closure#0}, rustc_middle[b86a10067019e542]::query::erase::Erased<[u8; 16usize]>>
  26:     0x7b71829df577 - <rustc_query_impl[ad104e1204eb17fd]::query_impl::resolver_for_lowering_raw::dynamic_query::{closure#2} as core[6fd456830ce7e084]::ops::function::FnOnce<(rustc_middle[b86a10067019e542]::ty::context::TyCtxt, ())>>::call_once
  27:     0x7b718369f198 - rustc_query_system[2707bc4b3550dbed]::query::plumbing::try_execute_query::<rustc_query_impl[ad104e1204eb17fd]::DynamicConfig<rustc_query_system[2707bc4b3550dbed]::query::caches::SingleCache<rustc_middle[b86a10067019e542]::query::erase::Erased<[u8; 16usize]>>, false, false, false>, rustc_query_impl[ad104e1204eb17fd]::plumbing::QueryCtxt, false>
  28:     0x7b718369ed07 - rustc_query_impl[ad104e1204eb17fd]::query_impl::resolver_for_lowering_raw::get_query_non_incr::__rust_end_short_backtrace
  29:     0x7b718373752d - rustc_interface[b107db532c72b558]::passes::create_and_enter_global_ctxt::<core[6fd456830ce7e084]::option::Option<rustc_interface[b107db532c72b558]::queries::Linker>, rustc_driver_impl[c7033648bfdf4c80]::run_compiler::{closure#0}::{closure#2}>::{closure#2}::{closure#0}
  30:     0x7b71836b26a0 - rustc_interface[b107db532c72b558]::interface::run_compiler::<(), rustc_driver_impl[c7033648bfdf4c80]::run_compiler::{closure#0}>::{closure#1}
  31:     0x7b718355afc8 - std[73fc39335cd1ceda]::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface[b107db532c72b558]::util::run_in_thread_with_globals<rustc_interface[b107db532c72b558]::util::run_in_thread_pool_with_globals<rustc_interface[b107db532c72b558]::interface::run_compiler<(), rustc_driver_impl[c7033648bfdf4c80]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>
  32:     0x7b718355b8b4 - <<std[73fc39335cd1ceda]::thread::Builder>::spawn_unchecked_<rustc_interface[b107db532c72b558]::util::run_in_thread_with_globals<rustc_interface[b107db532c72b558]::util::run_in_thread_pool_with_globals<rustc_interface[b107db532c72b558]::interface::run_compiler<(), rustc_driver_impl[c7033648bfdf4c80]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1} as core[6fd456830ce7e084]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  33:     0x7b718355ccab - std::sys::pal::unix::thread::Thread::new::thread_start::hbc7f68edabc6a613
  34:     0x7b717d6a370a - <unknown>
  35:     0x7b717d727aac - <unknown>
  36:                0x0 - <unknown>

error: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: rustc 1.87.0-nightly (351686bcf 2025-03-01) running on x86_64-unknown-linux-gnu

query stack during panic:
#0 [resolver_for_lowering_raw] getting the resolver for lowering
end of query stack
error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0658`.

```

</p>
</details>

<!--
query stack:
failed to reparse {mv_kind:?}: Some(DiagInner { level: Error, messages: [(FluentIdentifier("parse_expected_identifier_found_str", None), NoStyle)], code: None, span: MultiSpan { primary_spans: [/tmp/icemaker_global_tempdir.wxhLJkXfbxMZ/rustc_testrunner_tmpdir_reporting.FgtN60LYeTIA/mvce.rs:3:17: 3:28 (#0)], span_labels: [(/tmp/icemaker_global_tempdir.wxhLJkXfbxMZ/rustc_testrunner_tmpdir_reporting.FgtN60LYeTIA/mvce.rs:3:17: 3:28 (#0), FluentIdentifier("parse_expected_identifier", None))] }, children: [Subdiag { level: Help, messages: [(FluentIdentifier("parse_invalid_identifier_with_leading_number", None), NoStyle)], span: MultiSpan { primary_spans: [/tmp/icemaker_global_tempdir.wxhLJkXfbxMZ/rustc_testrunner_tmpdir_reporting.FgtN60LYeTIA/mvce.rs:3:17: 3:25 (#0)], span_labels: [] } }], suggestions: Enabled([]), args: {"token": Str("0b101010f64")}, sort_span: /tmp/icemaker_global_tempdir.wxhLJkXfbxMZ/rustc_testrunner_tmpdir_reporting.FgtN60LYeTIA/mvce.rs:3:17: 3:28 (#0), is_lint: None, long_ty_path: None, emitted_at: DiagLocation { file: "compiler/rustc_parse/src/parser/diagnostics.rs", line: 371, col: 34 } })
#0 [resolver_for_lowering_raw] getting the resolver for lowering

-->

--- a/compiler/rustc_ast_lowering/src/item.rs
+++ b/compiler/rustc_ast_lowering/src/item.rs
@@ -676,12 +676,9 @@ impl<'hir> LoweringContext<'_, 'hir> {
                 let ty =
                     self.lower_ty(ty, ImplTraitContext::Disallowed(ImplTraitPosition::StaticTy));
                 let safety = self.lower_safety(*safety, hir::Safety::Unsafe);
-
-                // njn: where for this?
                 if define_opaque.is_some() {
                     self.dcx().span_err(i.span, "foreign statics cannot define opaque types");
                 }
-
                 (ident, hir::ForeignItemKind::Static(ty, *mutability, safety))
             }
             ForeignItemKind::TyAlias(box TyAlias { ident, .. }) => {

--- a/compiler/rustc_parse/src/parser/mod.rs
+++ b/compiler/rustc_parse/src/parser/mod.rs
@@ -512,6 +512,14 @@ impl<'a> Parser<'a> {
         self
     }
 
+    #[inline]
+    fn with_recovery<T>(&mut self, recovery: Recovery, f: impl FnOnce(&mut Self) -> T) -> T {
+        let old = mem::replace(&mut self.recovery, recovery);
+        let res = f(self);
+        self.recovery = old;
+        res
+    }
+
     /// Whether the parser is allowed to recover from broken code.
     ///
     /// If this returns false, recovering broken code into valid code (especially if this recovery does lookahead)
@@ -770,7 +778,14 @@ impl<'a> Parser<'a> {
             && match_mv_kind(mv_kind)
         {
             self.bump();
-            let res = f(self).expect("failed to reparse {mv_kind:?}");
+
+            // Recovery is disabled when parsing macro arguments, so it must
+            // also be disabled when reparsing pasted macro arguments,
+            // otherwise we get inconsistent results (e.g. #137874).
+            let res = self.with_recovery(Recovery::Forbidden, |this| {
+                f(this).expect("failed to reparse {mv_kind:?}")
+            });
+
             if let token::CloseDelim(delim) = self.token.kind
                 && let Delimiter::Invisible(InvisibleOrigin::MetaVar(mv_kind)) = delim
                 && match_mv_kind(mv_kind)

--- a/tests/crashes/137874.rs
+++ b/tests/crashes/137874.rs
@@ -1,4 +1,0 @@
-//@ known-bug: #137874
-fn a() {
-    match b { deref !(0c) };
-}

--- a/tests/ui/associated-consts/issue-93835.rs
+++ b/tests/ui/associated-consts/issue-93835.rs
@@ -3,11 +3,10 @@
 fn e() {
     type_ascribe!(p, a<p:p<e=6>>);
     //~^ ERROR cannot find type `a` in this scope
-    //~| ERROR path separator must be a double colon
     //~| ERROR cannot find value
     //~| ERROR associated const equality
+    //~| ERROR cannot find trait `p` in this scope
     //~| ERROR associated const equality
-    //~| ERROR failed to resolve: use of unresolved module or unlinked crate `p`
 }
 
 fn main() {}

--- a/tests/ui/associated-consts/issue-93835.stderr
+++ b/tests/ui/associated-consts/issue-93835.stderr
@@ -1,15 +1,3 @@
-error: path separator must be a double colon
-  --> $DIR/issue-93835.rs:4:25
-   |
-LL |     type_ascribe!(p, a<p:p<e=6>>);
-   |                         ^
-   |
-   = note: if you meant to annotate an expression with a type, the type ascription syntax has been removed, see issue #101728 <https://github.com/rust-lang/rust/issues/101728>
-help: use a double colon instead
-   |
-LL |     type_ascribe!(p, a<p::p<e=6>>);
-   |                          +
-
 error[E0425]: cannot find value `p` in this scope
   --> $DIR/issue-93835.rs:4:19
    |
@@ -21,6 +9,12 @@ error[E0412]: cannot find type `a` in this scope
    |
 LL |     type_ascribe!(p, a<p:p<e=6>>);
    |                      ^ not found in this scope
+
+error[E0405]: cannot find trait `p` in this scope
+  --> $DIR/issue-93835.rs:4:26
+   |
+LL |     type_ascribe!(p, a<p:p<e=6>>);
+   |                          ^ not found in this scope
 
 error[E0658]: associated const equality is incomplete
   --> $DIR/issue-93835.rs:4:28
@@ -43,15 +37,7 @@ LL |     type_ascribe!(p, a<p:p<e=6>>);
    = note: this compiler was built on YYYY-MM-DD; consider upgrading it if it is out of date
    = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
 
-error[E0433]: failed to resolve: use of unresolved module or unlinked crate `p`
-  --> $DIR/issue-93835.rs:4:24
-   |
-LL |     type_ascribe!(p, a<p:p<e=6>>);
-   |                        ^ use of unresolved module or unlinked crate `p`
-   |
-   = help: you might be missing a crate named `p`
+error: aborting due to 5 previous errors
 
-error: aborting due to 6 previous errors
-
-Some errors have detailed explanations: E0412, E0425, E0433, E0658.
-For more information about an error, try `rustc --explain E0412`.
+Some errors have detailed explanations: E0405, E0412, E0425, E0658.
+For more information about an error, try `rustc --explain E0405`.

--- a/tests/ui/macros/failed-to-reparse-issue-137874.rs
+++ b/tests/ui/macros/failed-to-reparse-issue-137874.rs
@@ -1,0 +1,12 @@
+// This originally crashed because `Recovery::Forbidden` wasn't being applied
+// when fragments pasted by declarative macros were reparsed.
+
+macro_rules! m {
+    ($p:pat) => {
+        if let $p = 0 {}
+    }
+}
+
+fn main() {
+    m!(0X0); //~ ERROR invalid base prefix for number literal
+}

--- a/tests/ui/macros/failed-to-reparse-issue-137874.stderr
+++ b/tests/ui/macros/failed-to-reparse-issue-137874.stderr
@@ -1,0 +1,10 @@
+error: invalid base prefix for number literal
+  --> $DIR/failed-to-reparse-issue-137874.rs:11:8
+   |
+LL |     m!(0X0);
+   |        ^^^ help: try making the prefix lowercase (notice the capitalization): `0x0`
+   |
+   = note: base prefixes (`0xff`, `0b1010`, `0o755`) are lowercase
+
+error: aborting due to 1 previous error
+

